include/revng/Model/Binary.h

```diff
@@ -54,10 +56,9 @@ type: struct
     optional: true
   - name: DefaultPrototype
     doc: The default function prototype
-    reference:
-      pointeeType: TypeDefinition
-      rootType: Binary
+    type: Type
```

Comment:
1. Why switch to `Type`? Is this part of a wider effort of having only `Type`s around as opposed to `TupleTreeReference<TypeDefinition>`?
2. Do we have a check in `::verify` to restrict what this can be? Basically I think we want it to be a `DefinedType` referring a `RawFunctionDefinition`/`CABIFunctionDefinition`, correct?

Reply:
1. Raw references suck. They are only reasonable in a couple of places, so it felt like a waste to have to provide all the helper infrastructure just to support them (for example, we'd need to make a public way to extract reference for a newly create reference again). And the only downside is an extra line in the serialized form. I feel like after we have the default value  for model types in place, we should set the `DefinedType` to be the default anyway: which gets rid of this extra line.

2. Yep, it queries `isPrototype`.

Reply:
1. If you want to go this route, in `verify`, we have to ban `IsConst` (maybe) and ensure, there's no pointers, arrays. I don't want weird stuff here.

Reply:
> no pointers, arrays

`isPrototype` checks for that.

> `IsConst`

Actually this is interesting. But I don't think `verify` is the place. If we want to ban const prototypes, we should do so in the `isPrototype` itself.

Reply:
Mmmh, I disagree, either `const` is illegal in these kind of fields (that previously were `TupleTreeReference`s) or we simply allow users to put them and then basically ignore them.

I don't see why one should put `const` there, it's either a bug or a wrong expectation. So maybe it's better to ban it.

Reply:
I think there's a misunderstanding.

What I'm trying to say is that we have a `{as/is}Prototype` helper now, if we want to enforce some prototype specific semantics, I'd say we enforce them there.

This way, the verification of the field boils down to just "is `Prototype` field set to something that `isPrototype()` - good, otherwise - the model is invalid".

Reply:
Ah because `isPrototype` is used in verification?

Usually we tend to verify things at verify time and then assume they are like that. If you put the check in `isPrototype`, you're doing the check even out of validation, where we could assume that was the case.

Not too bad, but let's not do this too much. In general, we should verify things at verification time once, and then assume they hold.

Reply:
I feel like it's less about the verification and more about setting the semantics. By having `isPrototype` explicitly represent the thing we consider a `Prototype`, I feel like we're reducing the confusion about the terminology. It would be weird to, say, make a new CFT and set it to a function and then get confused why the "prototype" is fine, but the function is not if it's set.

> In general, we should verify things at verification time once, and then assume they hold.

Yep, but that sounds like a premature optimization 😉. Either way, I don't imagine an extra bool check per `isPrototype` PLUS replacing `skipConstAndTypedefs` with `skipConst` is going to make much difference.

Reply:
Shall we mark this as solved?